src/printer.js

```diff
@@ -1610,15 +1610,21 @@ function genericPrintNoParens(path, options, print) {
         return join(" & ", types);
       }
 
-      return group(
-        indent(
-          options.tabWidth,
-          concat([
-            ifBreak(concat([line, isIntersection ? "&" : "|", " "])),
-            join(concat([line, isIntersection ? "&" : "|", " "]), types)
-          ])
-        )
-      );
+      const parent = path.getParentNode();
+      // If there's a leading comment, the parent is doing the indentation
+      const shouldIndent = !(parent.type === 'TypeAlias' &&
+        hasLeadingOwnLineComment(options.originalText, n));
+
+      const token = isIntersection ? "&" : "|";
+      const printed = concat([
+        ifBreak(concat([shouldIndent ? line : "", token, " "])),
+        join(concat([line, token, " "]), types)
+      ]);
+
+      if (!shouldIndent) {
+        return group(printed);
+      }
+      return group(indent(options.tabWidth, printed));
```

Comment:
You could keep the code more like it was before and just do `indent(shouldIndent ? options.tabWidth 0, ...)`. I think it reads better when the doc is all one expression